swarm/api/http/server.go

```diff
@@ -81,33 +81,169 @@ type ServerConfig struct {
 	CorsString string
 }
 
-// browser API for registering bzz url scheme handlers:
-// https://developer.mozilla.org/en/docs/Web-based_protocol_handlers
-// electron (chromium) api for registering bzz url scheme handlers:
-// https://github.com/atom/electron/blob/master/docs/api/protocol.md
-
-// starts up http server
-func StartHTTPServer(api *api.API, config *ServerConfig) {
+func NewServer(api *api.API, corsString string) *Server {
 	var allowedOrigins []string
-	for _, domain := range strings.Split(config.CorsString, ",") {
+	for _, domain := range strings.Split(corsString, ",") {
 		allowedOrigins = append(allowedOrigins, strings.TrimSpace(domain))
 	}
 	c := cors.New(cors.Options{
 		AllowedOrigins: allowedOrigins,
-		AllowedMethods: []string{"POST", "GET", "DELETE", "PATCH", "PUT"},
+		AllowedMethods: []string{http.MethodPost, http.MethodGet, http.MethodDelete, http.MethodPatch, http.MethodPut},
 		MaxAge:         600,
 		AllowedHeaders: []string{"*"},
 	})
-	hdlr := c.Handler(NewServer(api))
 
-	go http.ListenAndServe(config.Addr, hdlr)
+	mux := http.NewServeMux()
+	server := &Server{api: api}
+	mux.HandleFunc("/bzz:/", server.WrapHandler(true, server.HandleBzz))
+	mux.HandleFunc("/bzz-raw:/", server.WrapHandler(true, server.HandleBzzRaw))
+	mux.HandleFunc("/bzz-immutable:/", server.WrapHandler(true, server.HandleBzzImmutable))
+	mux.HandleFunc("/bzz-hash:/", server.WrapHandler(true, server.HandleBzzHash))
+	mux.HandleFunc("/bzz-list:/", server.WrapHandler(true, server.HandleBzzList))
+	mux.HandleFunc("/bzz-resource:/", server.WrapHandler(true, server.HandleBzzResource))
+
+	mux.HandleFunc("/", server.WrapHandler(false, server.HandleRootPaths))
```

Comment:
Rather than a boolean argument to `WrapHandler`, I think the behaviour is clearer and more extendable with explicitly named middleware, for example:

```go
func (s *Server) Wrap(handler handlerFunc, middleware ...middlewareFunc) http.Handler {
  // ...
}

mux.HandleFunc("/bzz-list:/", s.Wrap(s.HandleBzzList, s.WithParseBzzURI))

mux.HandleFunc("/", s.Wrap(s.HandleRootPaths))
```

Reply:
yes, this is already in view, and i have (i think, at least), the correct solution to do [this](https://medium.com/@matryer/writing-middleware-in-golang-and-how-go-makes-it-so-much-fun-4375c1246e81), in combination with @janos's router code.

but i'd rather merge this part as-is at the moment, and cook a little bit longer which specific middleware components we'd need as part of the process of refactoring the other parts of the code (request parsing, context initialization, error handling, template parsing and rendering etc), then just submit this as the last PR (this is already in the game-plan in #770 - as PR 5)

Reply:
That Adapter pattern is pretty much what I'm proposing here, yes.